src/Sarif.Driver/Sdk/SkimmerBase.cs

```diff
@@ -18,37 +18,57 @@ public SkimmerBase()
 
         abstract public string Help { get; }
 
-        private IDictionary<string, string> messageFormats;
+        private IDictionary<string, string> messageTemplates;
+        private IDictionary<string, string> richMessageTemplates;
 
         abstract protected ResourceManager ResourceManager { get; }
 
-        abstract protected IEnumerable<string> FormatIds { get; }
+        abstract protected IEnumerable<string> TemplateIds { get; }
 
         virtual public RuleConfiguration Configuration {  get { return RuleConfiguration.Enabled; } }
 
         virtual public ResultLevel DefaultLevel { get { return ResultLevel.Warning; } }
 
-        virtual public IDictionary<string, string> MessageFormats
+        virtual public IDictionary<string, string> MessageTemplates
         {
             get
             {
-                if (this.messageFormats == null)
+                if (this.messageTemplates == null)
                 {
-                    this.messageFormats = InitializeMessageFormats();
+                    this.messageTemplates = InitializeMessageTemplates();
                 }
-                return this.messageFormats;
+                return this.messageTemplates;
             }
         }
 
-        private Dictionary<string, string> InitializeMessageFormats()
+        virtual public IDictionary<string, string> RichMessageTemplates
         {
-            return RuleUtilities.BuildDictionary(ResourceManager, FormatIds, Id);
+            get
+            {
+                if (this.richMessageTemplates == null)
+                {
+                    this.richMessageTemplates = InitializeRichMessageTemplates();
+                }
+                return this.richMessageTemplates;
+            }
+        }
+
+        private Dictionary<string, string> InitializeMessageTemplates()
+        {
+            return RuleUtilities.BuildDictionary(ResourceManager, TemplateIds, Id);
+        }
+
+        private Dictionary<string, string> InitializeRichMessageTemplates()
+        {
+            return new Dictionary<string, string>();
```

Comment:
This is the part that's not implemented. See the difference between this method and `InitializeMessageTemplates`. `RuleUtilities.BuildDictionary` relies on a naming convention for resource ids to to populate the `MessageTemplates` dictionary. We need to decide on a similar resource convention for populating the `RichMessageTemplates` dictionary. It might be as simple as `<rule id>_rich_<template id>`. For example, BinSkim has a resource id `BA2005_Error_Bad_Module`. Perhaps the rich text equivalent can be simply `BA2005_rich_Error_Bad_Module`.

Reply:
ok. can we get an issue opened on this?

---
In reply to: [161602933](https://github.com/Microsoft/sarif-sdk/pull/745#discussion_r161602933) [](ancestors = 161602933)

Reply:
@michaelcfanning I opened #746:

> In `SkimmerBase`, the `MessageTemplates` property is lazily initialized by a call to `InitializeMessageTemplates`, which calls `RuleUtilities.BuildDictionary` to populate the property by reading resource strings whose names follow a certain convention.

> At present, there is no similar mechanism for populating `SkimmerBase.RichMessageTemplates`; that property is always an empty dictionary. We need to define a convention for the resource strings which define rich message templates, then implement the logic (filling in the body of `SkimmerBase.InitializeRichMessageTemplates`, which currently does nothing), and write unit tests for it.